cpp/unreal_plugins/SimulationController/Source/SimulationController/VehicleAgent.cpp

```diff
@@ -0,0 +1,266 @@
+        SP_ASSERT(imu_sensor_);
+    }
+}
+
+VehicleAgent::~VehicleAgent()
+{
+    SP_LOG_CURRENT_FUNCTION();
+
+    auto observation_components = Config::get<std::vector<std::string>>("SIMULATION_CONTROLLER.VEHICLE_AGENT.OBSERVATION_COMPONENTS");
+
+    if (Std::contains(observation_components, "imu")) {
+        SP_ASSERT(imu_sensor_);
+        imu_sensor_ = nullptr;
+    }
+
+    if (Std::contains(observation_components, "camera")) {
+        SP_ASSERT(camera_sensor_);
+        camera_sensor_ = nullptr;
+    }
+
+    SP_ASSERT(vehicle_pawn_);
+    vehicle_pawn_->Destroy();
+    vehicle_pawn_ = nullptr;
+}
+
+void VehicleAgent::findObjectReferences(UWorld* world) {}
+
+void VehicleAgent::cleanUpObjectReferences() {}
+
+std::map<std::string, ArrayDesc> VehicleAgent::getActionSpace() const
+{
+    std::map<std::string, ArrayDesc> action_space;
+    auto action_components = Config::get<std::vector<std::string>>("SIMULATION_CONTROLLER.VEHICLE_AGENT.ACTION_COMPONENTS");
+
+    if (Std::contains(action_components, "set_drive_torques")) {
+        ArrayDesc array_desc;
+        array_desc.low_ = std::numeric_limits<float>::lowest();
+        array_desc.high_ = std::numeric_limits<float>::max();
+        array_desc.shape_ = { 4 };
+        array_desc.datatype_ = DataType::Float64;
+        action_space["set_drive_torques"] = std::move(array_desc);
+    }
+
+    if (Std::contains(action_components, "set_brake_torques")) {
+        ArrayDesc array_desc;
+        array_desc.low_ = std::numeric_limits<float>::lowest();
+        array_desc.high_ = std::numeric_limits<float>::max();
+        array_desc.shape_ = { 4 };
+        array_desc.datatype_ = DataType::Float64;
+        action_space["set_brake_torques"] = std::move(array_desc);
+    }
+
+    return action_space;
+}
+
+std::map<std::string, ArrayDesc> VehicleAgent::getObservationSpace() const
+{
+    std::map<std::string, ArrayDesc> observation_space;
+    auto observation_components = Config::get<std::vector<std::string>>("SIMULATION_CONTROLLER.VEHICLE_AGENT.OBSERVATION_COMPONENTS");
+
+    if (Std::contains(observation_components, "position")) {
+        ArrayDesc array_desc;
+        array_desc.low_ = std::numeric_limits<float>::lowest();
+        array_desc.high_ = std::numeric_limits<float>::max();
+        array_desc.datatype_ = DataType::Float64;
+        array_desc.shape_ = { 3 };
+        observation_space["position"] = std::move(array_desc); // position (X, Y, Z) in [cms] of the agent relative to the world frame.
+    }
+
+    if (Std::contains(observation_components, "orientation")) {
+        ArrayDesc array_desc;
+        array_desc.low_ = std::numeric_limits<float>::lowest();
+        array_desc.high_ = std::numeric_limits<float>::max();
+        array_desc.datatype_ = DataType::Float64;
+        array_desc.shape_ = { 3 };
+        observation_space["orientation"] = std::move(array_desc); // orientation (Roll, Pitch, Yaw) in [degs] of the agent relative to the world frame.
```

Comment:
comment is incorrect, it should say pyr because that is Unreal's convention and that is what the code actually does

Reply:
Corrected